### PR TITLE
[runtime] Stop swallowing string errors and work around environment races

### PR DESCRIPTION
**Pull request by alexanderkyte**
Fixes https://bugzilla.xamarin.com/show_bug.cgi?id=53461

### CODE REVIEW

File: mono/metadata/object.c [hunk 1]
```diff
@@ -6176,24 +6176,25 @@ mono_string_new_checked (MonoDomain *domain, const char *text, MonoError *error)
...
-    else
-        g_error_free (eg_error);
+	GError *eg_error = NULL;
+	MonoString *o = NULL;
+	guint16 *ut;
+	glong items_written;
+	int l;
 
-    g_free (ut);
+	mono_error_init (error);
```
Comment (lambdageek) [inline mono/metadata/object.c]: use `error_init (error)` - it'll port cleaner to master

**Comment by marek-safar** [comment]
I am not sure how this is related to #53461

**Comment by alexanderkyte** [comment]
Debugging this was a long story. 

So the icall that returns this value was returning null sometimes. This should have hit the assertions in the first commit, but they didn't because we were returning null but not setting the error. 

After fixing this, it was apparent that we were returning NULL because the string conversion had garbage characters in it. 

This happened despite the fact that the mono_cfg_dir was being set correctly. By tracing memory accesses, I found that we were altering the memory around that memory when we freed the memory allocation block and zeroed it out. 

I confirmed this was the case by observing that the mono_cfg_dir is null for most of an XI applications' execution after the free happens. I watched the glib internals assign that memory block to another malloc in the cases that we crashed, and the garbage memory was indeed in that region.

**Comment by alexanderkyte** [comment]
It was clear that something was wrong when the pointer that we sent to setenv and the pointer returned from g_getenv was exactly the same. In view of the fact that we free that pointer, we couldn't have anything but a bug.

**Comment by marek-safar** [comment]
Oh you spent more time on it already! I just started hit the problem at https://github.com/mono/mono/blob/master/mcs/class/System.Configuration/System.Configuration/ConfigurationFileMap.cs#L39 where the value is sometimes NULL which matches your findings

**Comment by alexanderkyte** [comment]
Yep, that's the code path.

File: mono/metadata/mono-config.c [hunk 2]
```diff
@@ -674,9 +674,11 @@ void
...
 {
 	/* If this variable is set, overrides the directory computed */
-	mono_cfg_dir = g_getenv ("MONO_CFG_DIR");
-	if (mono_cfg_dir == NULL)
-		mono_cfg_dir = mono_cfg_dir_allocated = g_strdup (dir);
+	const char *env_mono_cfg_dir = g_getenv ("MONO_CFG_DIR");
+	if (env_mono_cfg_dir == NULL)
+		env_mono_cfg_dir = dir;
+
+	mono_cfg_dir = mono_cfg_dir_allocated = g_strdup (env_mono_cfg_dir);
```
Comment (vargaz) [inline mono/metadata/mono-config.c]: Why do we need two variables which point at the same address ?

Comment (alexanderkyte) [inline mono/metadata/mono-config.c]: In the past only one of them was allocated, and should be freed. Let me remove the _allocated one.

**Comment by vargaz** [comment]
I think we should finish the review of this and commit to master before opening backport prs to other branches.

**Comment by spouliot** [comment]
I still have the same question from slack https://xamarinhq.slack.com/archives/C03CCJHCF/p1490219610173319
Copied here:

> so a decoding error will throw a `ExecutionEnvironmentException` so that’s not the fix
> 
> is this because we copy the `g_getenv` pointer ? and docs says it’s bad to do
> `The returned string may be overwritten by the next call to g_getenv(), g_setenv() or g_unsetenv().`
>

**Pull request closed without merging**